### ISSUE TITLE
New check: nested if/for

### ISSUE THREAD

**Issue by george-007** [checkstyle/checkstyle#5487]
There are ``NestedIfDepth`` and ``NestedForDepth`` checks.
It would be useful if one could also check **combined** nested depth for ``if/for`` (and maybe also for ``while/swith/try``). 

```java
for (int j = 0; j < 5; j++) {
    if (condition) {
        if (condition) {
            for (int i = 0; i < 5; i++) { 
               // 1 level of for nesting, 1 level of if nesting
               // but 3 levels of if/for nesting                    
            }
        }
    }
}
```

**Comment by romani** [comment]
@george-007 , let's imagine such Check implemented, please share imaginary configuration samples and code it should violate and pass.

**Comment by george-007** [comment]
Something similar to:
```java
package com.puppycrawl.tools.checkstyle.checks.coding;

import com.puppycrawl.tools.checkstyle.FileStatefulCheck;
import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
import com.puppycrawl.tools.checkstyle.api.DetailAST;
import com.puppycrawl.tools.checkstyle.api.TokenTypes;
import com.puppycrawl.tools.checkstyle.utils.CheckUtils;

/**
 * Check the number of nested {@code if,for,try,switch} statements. The maximum
 * number of nested layers can be configured. The default value is 2.
 * The code for the class is copied from the NestedIfDepthCheck-class.
 * The only difference are the intercepted tokens.
 * Example:
 * <pre>{@code
 * <!-- Restricts nested if, for, try, switch blocks to a specified depth (default = 2). -->
 * <module name="com.puppycrawl.tools.checkstyle.checks.coding.NestedCombinedDepthCheck">
 *   <property name="severity" value="info"/>
 *   <property name="max" value="2"/>
 * </module>}</pre>
 * @see NestedIfDepthCheck
 */
@FileStatefulCheck
public final class NestedCombinedDepthCheck extends AbstractCheck {

    /**
     * A key is pointing to the warning message text in "messages.properties"
     * file.
     */
    public static final String MSG_KEY = "nested.combined.depth";

    /** Maximum allowed nesting depth. */
    private int max = 2;
    /** Current nesting depth. */
    private int depth;

    /**
     * Setter for maximum allowed nesting depth.
     * @param max maximum allowed nesting depth.
     */
    public void setMax(int max) {
        this.max = max;
    }

    @Override
    public int[] getDefaultTokens() {
        return getRequiredTokens();
    }

    @Override
    public int[] getAcceptableTokens() {
        return getRequiredTokens();
    }

    @Override
    public int[] getRequiredTokens() {
        return new int[] {
            TokenTypes.LITERAL_IF,
            TokenTypes.LITERAL_FOR,
            TokenTypes.LITERAL_WHILE,
            TokenTypes.LITERAL_SWITCH,
            TokenTypes.LITERAL_TRY
        };
    }

    @Override
    public void beginTree(DetailAST rootAST) {
        depth = 0;
    }

    @Override
    public void visitToken(DetailAST ast) {
        if (!CheckUtils.isElseIf(ast)) {
            if (depth > max) {
                log(ast, MSG_KEY, depth, max);
            }
            ++depth;
        }
    }

    @Override
    public void leaveToken(DetailAST ast) {
        if (!CheckUtils.isElseIf(ast)) {
            --depth;
        }
    }

}

```

**Comment by romani** [comment]
@george-007 , implementation you provided is really weird and will not work on bunch of cases.

Let's focus on expected behavior over the code samples and xml configuration for them

**Comment by george-007** [comment]
Let say the configured maximum nesting depth is 2:
```java
boolean condition = true;
for (int j = 0; j < 5; j++) {
    if (condition) { // nesting + 1,  depth = 1, compliant
        if (condition) { // nesting + 1, depth = 2, compliant
            for (int i = 0; i < 5; i++) { // nesting + 1, depth = 3, violation
            }
        }
    }
}
```
```java
boolean condition = true;
try {
    if (condition) { // nesting + 1,  depth = 1, compliant
        if (condition) { // nesting + 1, depth = 2, compliant
            for (int i = 0; i < 5; i++) { // nesting + 1, depth = 3, violation
            }
        }
    }
} finally {
    // do nothing
}
```
```java
String string = "";
boolean condition = true;
switch (string) {
    case "a":
        for (int j = 0; j < 5; j++) { // nesting + 1,  depth = 1, compliant
            if (condition) {  // nesting + 1, depth = 2, compliant
                if (condition) { // nesting + 1, depth = 3, violation
                }
            }
        }
        break;
    default:
        break;
}
```

**Comment by romani** [comment]
Thanks a lot, did you try http://checkstyle.sourceforge.net/config_misc.html#DescendantToken ? I am not sure what tokens to search , but most likely SLIST.

Do you care about nesting curly braces or just block statements in general.
What behavior is expected on case code you provided, but without curly braces "{ }"

**Comment by rnveach** [comment]
@romani I asked you about something similar to this before in gitter many months ago.
`DescendantToken` can't be used because `else` is seen as a child of `if`.
So writing code like this `if (condition1) {} else if (condition2) {}` is seen as a depth of 2 instead of 1. This is why we have the custom check `NestedIfDepth` which has to override the behavior of `else-if`.
See https://github.com/checkstyle/checkstyle/blob/dabb75d43c7e02317565dde4c5e60f380d3b16b8/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NestedIfDepthCheck.java#L77

The way he wrote the code for the check `NestedCombinedDepthCheck` is similar to how I have it implemented and looks to be correct.

**Comment by romani** [comment]
>DescendantToken can't be used because else is seen as a child of if

Good.

But I still try him to think about Check design first.
I might miss smth , but I do not see how `(!CheckUtils.isElseIf(ast))` will do all magic in logic. I do not see DO_WHILE support and .... . I will not discuss implementation correctness, without clearly defined expected behavior. Such 3 code cases are not enough.

We are not a hosting to some not well designed Checks. It cost too much in maintenance.

Check should define his scope of code cases to validate, name of it should be properly selected, up to his behavior, .... As all agree on design, implementation is second step.

**Comment by Zlika** [comment]
Such feature would be very helpful, because currently you can trick the NestedIf/For rules by alterning if/for/while/... blocks.
